### PR TITLE
feat: EXPOSED-1023 Extend VECTOR data type to other supporting databases 

### PR DESCRIPTION
**Pull request by bog-walk**
#### Description

**Summary of the change**: Refactor current VECTOR data type support to cover as many databases as possible from the start. Refactor new column type to accept various dimensions formats, in what will hopefully be a way to avoid future complications if we add further features, like SPARSE and BINARY vectors.

**Detailed description**:
- **Why**: Oracle VECTOR support was kindly contributed in PR #2786 . To hit as many users as possible from the beginning and reduce future API breaking changes, the existing implementation needs to be ported over to the core and opened up. Particularly if we plan to support more dimensions formats, like BINARY. SPARSE vector types (Oracle + PostgreSQL) is a bit of a special case, and would most likely need its own subclass too --> simply because the input may not have a straightforward Kotlin input representation. Other ORMS occasionally provide a custom `SparseVectorObject` holder, which we should consider doing to; similar to `ExposedBlob`.

- **What**:
    - Extract majority of original `OracleVectorType` to new abstract class `BasicVectorType`
        - Add 2 subclasses currently, one that accepts `FloatArray` input and the other `IntArray`
        - Add some lines to support peculiarities of the other databases  
    -  Add `VectorFormat` enum class so that user can define dimensions format
    - Add vector distance function overloads, so user can also diff columns, like `mytable.embedding1.cosineDistance(mytable.embedding2)`
    - Add `vectorLiteral()` and `vectorParam()`
    - Move `Table.vector()` into a class method <-- keep original `Column<FloatArray>` since it seems the most common across the board
    - Add `Table.vector<type>()` that allows user to set underlying column input type +/- format options
    - Move `VectorDistance` class slightly out of namespace & generalize it to accept any expression types
        - It also now relies on dialect function provider + all supported overrides 
    - Add R2DBC `VectorTypeMapper` to handle edge cases when binding to statements 
    - Add documentation -> KDocs + website page for new type
    - Update tests:
        - Add migration tests <-- change in type/format/dimension not supported for same reasons as arrays
        - Port existing tests to R2DBC suite
        - Add some extra tests: DAO, default values, plain insert/select 

---

#### Type of Change

Please mark the relevant options with an "X":
- [X] New feature

Affected databases:
- [X] Oracle
- [X] MariaDB
- [X] Postgres
- [X] SqlServer

#### Checklist

- [X] Unit tests are in place
- [X] The build is green (including the Detekt check)
- [X] All public methods affected by my PR has up to date API docs
- [X] Documentation for my change is up to date

---

#### Related Issues
[EXPOSED-1023](https://youtrack.jetbrains.com/issue/EXPOSED-1023/Extend-VECTOR-data-type-to-other-supporting-databases)

### CODE REVIEW

File: buildScripts/docker/docker-compose-postgres.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 services:
     postgres:
-        image: postgres
+        image: pgvector/pgvector:pg18
```
Comment (bog-walk) [inline buildScripts/docker/docker-compose-postgres.yml:3]: Necessary to access PG vector extension. Should not impact rest of codebase tests as it sits on the official PG docker.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt [hunk 2]
```diff
@@ -1405,6 +1405,291 @@ class ArrayColumnType<T, R : List<Any?>>(
...
+    FLOAT64,
+
+    /** Each dimension is formatted as 8-bit integers. */
+    INT8,
+}
+
+/**
+ * Base column for storing vector values in various dimension formats.
+ */
+abstract class BasicVectorColumnType<T>(
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt:1429]: Opted for an abstract class approach to leave room for potential future additions to `VectorFormat`. At minimum, `FLOAT16` (SQL Server + PG) can be introduced using existing subclass.

`BINARY` for Oracle would also be good, just need to consider what Kotlin client representation to use. Since it's unsigned 8-bit integers, maybe `UByteArray` (experimental), otherwise straight `ByteArray` 🤔

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt [hunk 3]
```diff
@@ -1405,6 +1405,291 @@ class ArrayColumnType<T, R : List<Any?>>(
...
+    }
+}
+
+/**
+ * Column for storing vector values as an `IntArray`.
+ */
+class IntVectorColumnType(
+    dimensions: Int,
+    format: VectorFormat,
+) : BasicVectorColumnType<IntArray>(dimensions, format) {
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt:1541]: Other than for `BINARY`, `SPARSE` vectors might be the only other subclass to consider in the future.
The input format doesn't have a straightforward Kotlin representation, unless we force String:
- Oracle --> `INSERT INTO my_sparse_tab VALUES('[5,[2,4],[10,20]]')`
- PG --> `INSERT INTO documents (embedding) VALUES ('{1:0.5, 100:1.2, 500:0.8}/30522');`

I've seen other libraries provide their own custom objects, which might be the best option for us, like we did for `ExposedBlob`.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt [hunk 4]
```diff
@@ -1405,6 +1405,291 @@ class ArrayColumnType<T, R : List<Any?>>(
...
+    repeat(capacity) { i ->
+        val base = i * 4
+        val bits = (this[base + 3].toInt() and 0xFF shl 24) or
+            (this[base + 2].toInt() and 0xFF shl 16) or
+            (this[base + 1].toInt() and 0xFF shl 8) or
+            (this[base].toInt() and 0xFF)
+        result[i] = Float.fromBits(bits)
+    }
+    return result
+}
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt:1607]: I put all these here, because it's easier to review, but I believe they could be moved out of the way to an internal package too.

These exist primarily for MySQL/MariaDB because they return raw binary data when their vectors are queried. What an end-user should do is wrap the column in a function when calling select, like `SELECT VECTOR_TO_STRING(embedding)`.
This is the same problem we ran into with JSON + SQLite and I didn't want to "magically" wrap/change DSL queries just so the user can get human-readable content and use DAO without extra steps.
So I created these conversion functions instead. They are ugly but I didn't want to add more `ByteBuffer` either.
And they work.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt [hunk 5]
```diff
@@ -1405,6 +1405,291 @@ class ArrayColumnType<T, R : List<Any?>>(
...
+private object OracleVectorBinder {
+    private val vectorClass by lazy { Class.forName("oracle.sql.VECTOR") }
+    private val ofFloat32Values by lazy { vectorClass.getMethod("ofFloat32Values", Any::class.java) }
+    private val ofFloat64Values by lazy { vectorClass.getMethod("ofFloat64Values", Any::class.java) }
+    private val ofInt8Values by lazy { vectorClass.getMethod("ofInt8Values", Any::class.java) }
+
+    fun valueToBind(value: Any, format: VectorFormat): Any = when (format) {
+        VectorFormat.FLOAT32 -> ofFloat32Values.invoke(null, value)
+        VectorFormat.FLOAT64 -> ofFloat64Values.invoke(null, value)
+        VectorFormat.INT8 -> ofInt8Values.invoke(null, value)
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt:1676]: Moved over from the original implementation and expanded to cover more dimension formats.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/FunctionBuilder.kt [hunk 6]
```diff
@@ -222,3 +224,33 @@ infix operator fun <E, T : List<E>?> ExpressionWithColumnType<T>.get(index: Int)
...
+
+/** Returns the mathematical distance calculated between the two vector expressions, based on euclidean similarity. */
+fun <T> ExpressionWithColumnType<T>.euclideanDistance(target: Expression<T>): VectorDistance<T> =
+    VectorDistance(this, target, VectorDistanceMetric.EUCLIDEAN)
+
+/** Returns the mathematical distance calculated between this vector expression and a target vector, based on euclidean similarity. */
+fun ExpressionWithColumnType<FloatArray>.euclideanDistance(target: FloatArray): VectorDistance<FloatArray> = euclideanDistance(vectorParam(target))
+
+/** Returns the mathematical distance calculated between this vector expression and a target vector, based on euclidean similarity. */
+fun ExpressionWithColumnType<IntArray>.euclideanDistance(target: IntArray): VectorDistance<IntArray> = euclideanDistance(vectorParam(target))
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/FunctionBuilder.kt:246]: I'm not a fan of how much function duplication this approach needs, but it felt necessary to add overloads for `column vs column` and `column vs primitive`.
I tried to combine the bottom 2 into a single function, but all the variants of `vectorParam()` I tried ended up complaining (generics, inline reified).

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt [hunk 7]
```diff
@@ -1097,6 +1097,41 @@ open class Table(name: String = "") : ColumnSet(), DdlAware {
...
+     * Creates a vector column, with the specified [name], for storing elements of a `FloatArray`.
+     * By default, this stores vectors with each dimension formatted as 32-bit floating-point numbers.
+     *
+     * **Note** This column type is not supported by all databases and drivers. Please check your documentation.
+     *
+     * @param name Name of the column.
+     * @param dimensions The number of dimensions that each stored vector must have.
+     */
+    fun Table.vector(name: String, dimensions: Int): Column<FloatArray> =
+        registerColumn(name, FloatVectorColumnType(dimensions, VectorFormat.FLOAT32))
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt]: This is the original proposed column type, and from my research, it may be the most commonly reached for anyway. So I've left it as a convenience method. Then, if the user wants more explicit control over formatting options, they can go for the option below.

File: exposed-migration-jdbc/src/test/kotlin/org/jetbrains/exposed/v1/migration/jdbc/DatabaseMigrationTests.kt [hunk 8]
```diff
@@ -576,4 +576,45 @@ class DatabaseMigrationTests : DatabaseTestsBase() {
...
+    fun testVectorColumnTriggersMigration() {
+        val original = object : Table("vector_items") {
+            val id = integer("id")
+        }
+        val updated = object : Table("vector_items") {
+            val id = integer("id")
+            val embedding = vector("embedding", dimensions = 5)
+        }
+
+        withDb(vectorTypeSupportedDb - TestDB.SQLSERVER) { testDb ->
```
Comment (bog-walk) [inline exposed-migration-jdbc/src/test/kotlin/org/jetbrains/exposed/v1/migration/jdbc/DatabaseMigrationTests.kt:592]: SQL Server temporarily excluded because JDBC metadata does not return vector columns. It was a known issue [[here](https://github.com/microsoft/mssql-jdbc/pull/2883), [here](https://github.com/microsoft/mssql-jdbc/issues/2862)] and has been fixed. We just need to bump our MSSQL driver first.

File: exposed-r2dbc/src/main/kotlin/org/jetbrains/exposed/v1/r2dbc/mappers/VectorTypeMapper.kt [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import org.jetbrains.exposed.v1.core.vendors.MysqlDialect
+import org.jetbrains.exposed.v1.core.vendors.PostgreSQLDialect
+import kotlin.reflect.KClass
+
+/**
+ * Mapper for vector data types.
+ */
+class VectorTypeMapper : TypeMapper {
+    @Suppress("MagicNumber")
+    override val priority = 0.2
```
Comment (bog-walk) [inline exposed-r2dbc/src/main/kotlin/org/jetbrains/exposed/v1/r2dbc/mappers/VectorTypeMapper.kt:19]: Opted for the same priority level as `ArrayTypeMapper`. Let me know if anyone disagrees.

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/v1/tests/shared/types/VectorColumnTypeTests.kt [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+import kotlin.math.sqrt
+import kotlin.test.assertContentEquals
+import kotlin.test.assertTrue
+
+class VectorColumnTypeTests : DatabaseTestsBase() {
+    // Only MySQL9+ supports VECTOR, so it is excluded until we add this version to tests;
+    // Tested locally on Docker image mysql:9.7
+    private val vectorTypeSupportedDb = setOf(TestDB.ORACLE, TestDB.MARIADB, TestDB.POSTGRESQL, TestDB.SQLSERVER)
+    // When MySQL9 is added, it will need to be excluded from all VECTOR_DISTANCE tests, because this function
+    // is available only for users of HeatWave MySQL on OCI; it is not included in MySQL Commercial or Community distributions.
```
Comment (bog-walk) [inline exposed-tests/src/test/kotlin/org/jetbrains/exposed/v1/tests/shared/types/VectorColumnTypeTests.kt:34]: To confirm, MySQL should technically be on this list, but VECTOR is only available since 9.+ and we are still testing on 5.+/8.+. There are already tickets open to add 9.+ +/- drop MySQL5, we just haven't made plans for it. So temporarily I tested locally on a new image.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt [hunk 11]
```diff
@@ -1405,6 +1405,291 @@ class ArrayColumnType<T, R : List<Any?>>(
...
+    return result
+}
+
+@Suppress("MagicNumber")
+private fun ByteArray.toIntArray(): IntArray {
+    val capacity = this.size / 4
+    val result = IntArray(capacity)
+    repeat(capacity) { i ->
+        val base = i * 4
+        result[i] = (this[base].toInt() and 0xFF shl 24) or
```
Comment (obabichevjb) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/ColumnType.kt:1615]: I expected that it works if it's tested (even simple write/read value would fail if it's real error), but I'm surprised that `toFloatArray()` and `toIntArray()` fill `result` in different order (`this[base].toInt() and 0xFF shl 24` vs `this[base + 3].toInt() and 0xFF shl 24`, and other), so hope it's intended.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt [hunk 12]
```diff
@@ -1097,6 +1097,41 @@ open class Table(name: String = "") : ColumnSet(), DdlAware {
...
+    /**
+     * Creates a vector column, with the specified [name], for storing elements of a `FloatArray`.
+     * By default, this stores vectors with each dimension formatted as 32-bit floating-point numbers.
+     *
+     * **Note** This column type is not supported by all databases and drivers. Please check your documentation.
+     *
+     * @param name Name of the column.
+     * @param dimensions The number of dimensions that each stored vector must have.
+     */
+    fun Table.vector(name: String, dimensions: Int): Column<FloatArray> =
```
Comment (obabichevjb) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt]: Could be dimensions nullable? I see that postgres can use variable length vectors, but it requires to use `halfvec` (?), so porbably it's another column type. 

Found here [The Ultimate Guide to using PGVector](https://medium.com/@intuitivedl/the-ultimate-guide-to-using-pgvector-76239864bbfb) : `If you need variable-length vectors, use HALFVEC without specifying a dimension`

**Review by obabichevjb — APPROVED** [review]

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt (hunk 12 above)
Comment (bog-walk) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/v1/core/Table.kt]: Yeah, Oracle, PG, and MySQL allow undefined dimensions just by omitting it -> `VECTOR`. Added this functionality back in.
